### PR TITLE
feat: Add web query plugin

### PR DESCRIPTION
**Pull request by scaryrawr**
- Adds support for launching a web query from launcher

By prefix:
image

Also by if no results found:
image

### CODE REVIEW

**Comment by scaryrawr** [comment]
Latest commit adjusts description and name:
image

File: schemas/org.gnome.shell.extensions.pop-shell.gschema.xml [hunk 1]
```diff
@@ -185,8 +185,13 @@
             <summary>Swap window right</summary>
         </key>
 
-        <!-- Workspace Management -->
+        <!-- Launcher Options -->
+        <key type="s" name="search-engine">
```
Comment (mmstick) [inline schemas/org.gnome.shell.extensions.pop-shell.gschema.xml:189]: I think we should keep plugin settings out of the Pop Shell's settings

File: src/dialog_launcher.ts [hunk 2]
```diff
@@ -141,6 +141,33 @@ export class Launcher extends search.Search {
             // Truncate excess items from the list
             this.options.splice(this.list_max());
 
+            // Fall back to web query if nothing found
```
Comment (mmstick) [inline src/dialog_launcher.ts:144]: Maybe we should have a search-not-found condition for plugins?

File: src/prefs.ts [hunk 3]
```diff
@@ -17,8 +17,16 @@ interface AppWidgets {
     smart_gaps: any,
     snap_to_grid: any,
     window_titles: any,
+    search_engine: any,
 }
 
+const SEARCH_ENGINES: string[] = [
```
Comment (mmstick) [inline src/prefs.ts:23]: What if the search engines are based on the first word in the query?

wiki article
bing search
ddg search
google search

File: src/dialog_launcher.ts (hunk 2 above)
Comment (scaryrawr) [reply src/dialog_launcher.ts:144]: Hmmm... I wonder what a good name for a setting would be, kind of the "Default Prefix if no results found".

File: src/prefs.ts (hunk 3 above)
Comment (scaryrawr) [reply src/prefs.ts:23]: This makes sense, and a good way to remove it from the settings panel

File: src/dialog_launcher.ts (hunk 2 above)
Comment (scaryrawr) [reply src/dialog_launcher.ts]: I've removed this for now since it's a bigger behavior change.

**Comment by jacobgkau** [comment]
The Wikipedia option currently doesn't search Wikipedia, it only opens Wikipedia to the specified page. For example, when I type `wiki: test search`, I would expect to go to `https://en.wikipedia.org/w/index.php?search=test+search` for a list of search results, but instead I go to `https://en.wikipedia.org/wiki/Test_search`, which does not exist. (Vising the `?search=` link with the name of an existing page does redirect to the page.) Should the Wikipedia part of this plugin use a search URL instead of the article URL, or is it only meant to be used with existing pages?

For comparison, searching DuckDuckGo with `!w test search` goes to `https://en.wikipedia.org/wiki/Special:Search?search=test%20search&go=Go&ns0=1`. (This is a different URL than what I get when using Wikipedia's own search box, but appears to be the same page.)

**Comment by scaryrawr** [comment]
> The Wikipedia option currently doesn't search Wikipedia, it only opens Wikipedia to the specified page. For example, when I type `wiki: test search`, I would expect to go to `https://en.wikipedia.org/w/index.php?search=test+search` for a list of search results, but instead I go to `https://en.wikipedia.org/wiki/Test_search`, which does not exist. (Vising the `?search=` link with the name of an existing page does redirect to the page.) Should the Wikipedia part of this plugin use a search URL instead of the article URL, or is it only meant to be used with existing pages?
> 
> For comparison, searching DuckDuckGo with `!w test search` goes to `https://en.wikipedia.org/wiki/Special:Search?search=test%20search&go=Go&ns0=1`. (This is a different URL than what I get when using Wikipedia's own search box, but appears to be the same page.)

I've updated to use the `https://en.wikipedia.org/w/index.php?search=` as the base for wiki

**Review by jacobgkau — APPROVED** [review]

**Comment by mmstick** [comment]
I will make one change: `:` shouldn't be necessary to identify a web query. `wiki test search` should be fine

**Pull request merged**